Mailgun Integration

bpresser123 opened the pull request:
@robmadden I started integrating Mailgun into my app. I have a big portion of this complete but am not sure how to best approach the ```create``` method in my [incoming_controller](https://github.com/bpresser123/blocmarks/compare/mailgun#diff-586b07a8c9c0d78f29eb2e6d2bbfe22c) and if there is anything else that I need to adjust to make it work properly. I'm assuming I should be able to message postmaster@app92689d5e9ca74c018e434af8f4f3d120.mailgun.org with a subject and link to save a bookmark to my account. 

Using ```rails generate scaffold``` as directed, I added a [model](https://github.com/bpresser123/blocmarks/compare/mailgun#diff-072e18d79af30929353d6d6a6fc8ff60) and [controller](https://github.com/bpresser123/blocmarks/compare/mailgun#diff-77d760879b2a70798ada82a94332ca54) for the ```User``` functions. I used mguser (mailgun user) for the controller and model name because ```User``` is already a model in my app. Should I have done that differently?

robmadden commented:
I've never used Mailgun, but have you taken a look at [Bloc's resource](https://www.bloc.io/resources/mailgun-integration) on it that was listed in the checkpoint?

I'd read that and if still doesn't make sense, let's assess. Typically the best way to flush out an implementation of a method (i.e. your `incoming#create` is to write a test first describing what you want it to do. This technique also allows you to specify the exact input it should get.

It appears from Bloc's resource that they give instructions on how to test it locally. You can use that as a "bootstrap" to see how the input is formatted when it gets to `incoming#create`. Let me know if that helps.

bpresser123 commented:
@robmadden I was able to fix the ```redirect_to``` issue with the ```update``` method in my Bookmarks Controller. This commit has all the latest updates for Mailgun integration. Let me know if you find any good resources or solutions for implementing this mail feature and I will keep you posted on what I come up with too. Thanks!

robmadden commented:
I was able to get it to work following these links and using the debugger in RubyMine. I will submit my work so you can view what I did. Also, you'll need to edit the "Authorized Recipients " setting in MailGun's profile for your domain if it's not public.

bpresser123 commented:
Ok cool. Thanks!

bpresser123 commented:
@robmadden So I read through the links you sent over, other documentation, and changed some of the settings in my app and in Mailgun but am still struggling to get this to work. Can you point out what I'm missing? Do I just send an email to my postmaster address to test the app? I am still not sure about how to properly configure my ```incoming_controller``` for parsing incoming info either. Overall, The concept makes sense, it's just the specific setup that is kind of confusing to me.

robmadden commented:
Like I said, the code that I pushed up has everything it needs in order to
send a welcome email. You just need to emulate that code with your own
mailgun settings. As for incoming, it should "just work" if you reply to
the email as well but I didn't test it.

On Fri, Dec 16, 2016 at 2:20 PM bp123 <notifications@github.com> wrote:

> @robmadden <https://github.com/robmadden> So I read through the links you
> sent over, other documentation, and changed some of the settings in my app
> and in Mailgun but am still struggling to get this to work. Can you point
> out what I'm missing? Do I just send an email to my postmaster address to
> test the app? I am still not sure about how to properly configure my
> incoming_controller for parsing incoming info either. Overall, The
> concept makes sense, it's just the specific setup that is kind of confusing
> to me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bpresser123/blocmarks/pull/4#issuecomment-267710087>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcYb5hWyV0d1vwC0Be970eJleXou9afks5rIw65gaJpZM4LIU7s>
> .
>

bpresser123 commented:
I made all of the changes you suggested but still do not receive a welcome email when I send an email to postmaster and no new bookmarks or topics are posted to my account. My Mailgun Dashboard shows incoming emails so something must be making to my app but not returning any results to me as a user. In regards to replying to the email, the project says that anything I send over should be added as a new bookmark or topic depending on what info I add but doesn't talk about any functions relating to a response to an email.

robmadden commented:
Did you check your settings on Mailgun's website about allowing it to send to your email address? With private accounts, they restrict you from sending an email unless you say otherwise. If that isn't it, push your branch and I'll take a gander.

bpresser123 commented:
I added myself as an Authorized Recipient but still got no results. Here's the [Mailgun Integration](https://github.com/bpresser123/blocmarks/pull/4/commits/316f51db79e6d2f5d47ea75f4fc5f06bc45df7f7)  commit I added. Thanks!

bpresser123 commented:
@robmadden I made changes to my ```registrations_controller``` which is included in my [Registrations Controller Update](https://github.com/bpresser123/blocmarks/pull/4/commits/fdf71b820ce2eb498168d8c16b86636013acd7e6) commit. I tried using the debugger to test my code but wasn't getting any results. I watched a few tutorials on RubyMine but am not sure if I'm using the debugger correctly. I still don't get any results when sending info to my app. I've followed all of the directions I've been given but nothing seems to work. Can you see what I'm missing? I feel like I've added everything I need to use Mailgun...and again, I'm not sure about how to finish my ```incoming_controller```. Thanks again for the help!

robmadden commented:
Let's review next session.

On Sat, Dec 17, 2016 at 12:05 PM bp123 <notifications@github.com> wrote:

> @robmadden <https://github.com/robmadden> I made changes to my
> registrations_controller which is included in my Registrations Controller
> Update
> <https://github.com/bpresser123/blocmarks/pull/4/commits/fdf71b820ce2eb498168d8c16b86636013acd7e6>
> commit. I tried using the debugger to test my code but wasn't getting any
> results. I watched a few tutorials on RubyMine but am not sure if I'm using
> the debugger correctly. I still don't get any results when sending info to
> my app. I've followed all of the directions I've been given but nothing
> seems to work. Can you see what I'm missing? I feel like I've added
> everything I need to use Mailgun...and again, I'm not sure about how to
> finish my incoming_controller. Thanks again for the help!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bpresser123/blocmarks/pull/4#issuecomment-267783802>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcYb5H6oYaKPwtPlrrK_dlRwpJQOqonks5rJEBxgaJpZM4LIU7s>
> .
>

bpresser123 commented:
@robmadden I added some updates to the incoming_controller in my latest [commit](https://github.com/bpresser123/blocmarks/pull/4/commits/ee8206cb66745e977fc227784e83db041b78e1cb) and moved the spec code that we worked on over to my Mailgun branch. 

Since this concept is new, I am still not sure about how to parse incoming info. I think you sent me some links but I can't get any info from our last Screenhero session once the session is closed. I feel like most of the Mailgun configurations are complete but I still need to be able to handle incoming info and confirm user info. 

Does the ```create``` method in my ```incoming_controller``` need to redirect to the Devise ```registrations_contoller``` if no user info matches my database? If so, how would I bring it back to my ```incoming_controller``` to finish parsing the other info received from an email? Should I be reassigning any data to new params to be passed to my conditionals (:sender, :subject, "body-plain", etc)?  Any info or advice helps. Thanks!

robmadden commented:
> I think you sent me some links but I can't get any info from our last Screenhero session once the session is closed.

Feel free to record the sessions.

> I feel like most of the Mailgun configurations are complete but I still need to be able to handle incoming info and confirm user info.

Next session we will figure out how to drive this using TDD.

> Does the create method in my incoming_controller need to redirect to the Devise registrations_contoller if no user info matches my database?

This would probably be the easiest solution. I would just google on how to register a new user with devise.

> If so, how would I bring it back to my incoming_controller to finish parsing the other info received from an email?

When you make a call to a function, the program function call stack makes sure the execution will return to the callee function ensuring the remaining logic is evaluated.

> Should I be reassigning any data to new params to be passed to my conditionals (:sender, :subject, "body-plain", etc)? Any info or advice helps. Thanks!

If you want to change any of these values, then yes.

bpresser123 commented:
@robmadden I pushed all the current code for blocmarks in the [RSpec testing and debugging](https://github.com/bpresser123/blocmarks/pull/4/commits/bab7db032582227251590988f1803ab31299be6e) commit. Let me know what you figure out and I will continue to work on things as well. Thanks!

bpresser123 commented on spec/controllers/incoming_controller_spec.rb:
```diff
@@ -8,18 +8,26 @@
     it "parses an email correctly" do
-      expect_response = :email
-      post :create, "content"
+      expect_response = params[:sender, :subject, 'body-plain']
```
Not sure if this is how I should be confirming a response for parsing an email. The rest of the test makes sense to me overall.

bpresser123 commented:
@robmadden I made some changes to the `incoming_controller` and spec file. From my understanding, the tests seem to cover most of the `incoming_controller`'s functionality. I did a lot of research on RSpec and found a lot of helpful info in this [tutorial](https://www.tutorialspoint.com/rspec/rspec_basic_syntax.htm). 

I'm not sure if I'm handling the parsed info correctly on line 29 of the [spec file](https://github.com/bpresser123/blocmarks/pull/4/files/bab7db032582227251590988f1803ab31299be6e..6a95f2351a2340cabaae4ae556fab04d839e2fa1#diff-3a98c7fc155481876a49f0e961ab8a5c). The tests make a lot more sense to me now but I'm not sure what else needs to be written. Should I divide each piece of parsed info into it's own test? If so, will the tests even pass without the proper settings in my `incoming_controller`? I guess that's where I'm having the most trouble.

I also noticed that my app will send a welcome email to me when I try to sign up with info that already exists as an account. How can I stop that from happening? Also, is there a way to add a body to that message as well? I know the `user_mailer` file has associated views but I'm not sure about how to implement that into the code I have in my [user_mailer.rb](https://github.com/bpresser123/blocmarks/blob/mailgun/app/mailers/user_mailer.rb) file.

bpresser123 commented:
@robmadden Here are my spec errors:


  1) IncomingController POST #create returns http redirect if unauthenticated
     Failure/Error: @topic = Topic.find(params[:subject])
     
     ActiveRecord::RecordNotFound:
       Couldn't find Topic with 'id'=

***How can I find an id for my topic? ^***

  2) IncomingController POST #create returns http success if authenticated
     Failure/Error: @topic = Topic.find(params[:subject])
     
     ActiveRecord::RecordNotFound:
       Couldn't find Topic with 'id'=

  3) IncomingController POST #create parses an email correctly
     Failure/Error: expect_response = params
     
     NameError:
       undefined local variable or method `params' for #<RSpec::ExampleGroups::IncomingController::POSTCreate:0x007fc725564408>

***How can I set the values for params? ^***

  4) IncomingController POST #create creates a bookmark
     Failure/Error: @topic = Topic.find(params[:subject])
     
     ActiveRecord::RecordNotFound:
       Couldn't find Topic with 'id'=

  5) IncomingController POST #create creates a bookmark with expected value
     Failure/Error: @topic = Topic.find(params[:subject])
     
     ActiveRecord::RecordNotFound:
       Couldn't find Topic with 'id'=

Finished in 0.94507 seconds (files took 6.48 seconds to load)
5 examples, 5 failures

bpresser123 commented:
@robmadden I pushed a lot of commits to GitHub and Heroku to test out a few things with my `incoming_controller` updates. I added some lines of code to the [it "finds or creates user by params[:sender]"](https://github.com/bpresser123/blocmarks/pull/4/commits/d1bafe71e46067c863804018d267304509adca09#diff-3a98c7fc155481876a49f0e961ab8a5c) block for creating a bookmark considering it already covered 2/3 of the desired functions. My portion of the test fails but is headed in the right direction. Here's my error:

```bash
Failures:

  1) IncomingController POST #create (unauthenticated) finds or creates user by params[:sender]
     Failure/Error: expect(assigns("body-plain")).to eq bookmark.url
     
       expected: "Bookmark Content"
            got: nil
     
       (compared using ==)
     # ./spec/controllers/incoming_controller_spec.rb:40:in `block (3 levels) in <top (required)>'
```

How can I get "Bookmark Content" to return instead of `nil`? I added `@bookmark = Bookmark.new(url: params["body-plain"])` to the [create](https://github.com/bpresser123/blocmarks/pull/4/commits/d1bafe71e46067c863804018d267304509adca09#diff-586b07a8c9c0d78f29eb2e6d2bbfe22c) method in the `incoming_controller` but am not sure if there is another way to approach creating the new bookmark. The good news is that the app finds me as a user and creates a new topic when receiving an email...just no bookmark yet.

bpresser123 commented:
@robmadden Thanks for the info! I fully understand this spec and all of the tests pass. Now I just need to complete the bookmark portion of the create method. Why doesn't `.new` or `find_or_create_by` work for `@bookmark`?

robmadden commented:
Not sure. Try loading it up in pry or a debugger to inspect the model. Make
sure the model inherits from the proper classes i.e. ActionController::Base

On Tue, Jan 24, 2017 at 12:06 PM bp123 <notifications@github.com> wrote:

> @robmadden <https://github.com/robmadden> Thanks for the info! I fully
> understand this spec and all of the tests pass. Now I just need to complete
> the bookmark portion of the create method. Why doesn't .new or
> find_or_create_by work for @bookmark?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bpresser123/blocmarks/pull/4#issuecomment-274922821>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcYb6JFQz4FRZgDMEiumEYf8BduePK5ks5rVlm5gaJpZM4LIU7s>
> .
>
